proposals/0000-backpack.rst

```diff
@@ -0,0 +1,1030 @@
+   a dummy proxy type.
+
+2. In a multi-parameter type class, some methods may not mention
+   all of the types in the type class: these methods cannot be
+   resolved unambiguously without introducing a functional dependency.
+   Similarly, every type parameter must be listed in the constraints,
+   even if they are not used.  If there is a natural type to
+   modularize over, these issues can be alleviated by introducing
+   an *associated type*.
+
+3. And of course, you still have to put the damn constraint on
+   every parametrized function.
+
+**Preprocessor.**  A classic way to swap out a backing implementation
+is to replace it textually using a pre-processor.  In such a scheme, you
+might write your Haskell program like this::
+
+    import Data.ByteString as Str
+
+    #include "Impl.hs"
+
+Besides the intrinsic ugliness of such an approach, it fails to define
+the *interface* between an implementation and the backing
+implementations it relies on.  Without this type information, separate
+typechecking is impossible.  Separate typechecking is extremely
+useful: otherwise, you get extremely difficult to understand error
+messages, as is seen in C++ template error messages.
+
+Overview
+~~~~~~~~
+
+In this section, we give a high-level overview of Backpack, using the
+string-processing problem as our running example. In this section, we
+will introduce important terminology (**bolded**) that we will be used
+in the rest of this specification.
+
+Backpack solves this string-processing problem in the following way:
+To parametrize over a string implementation, a user of Backpack writes a
+**signature** describing the necessary supported interface for their
+strings  (an *egalitarian interface.*) A signature is much like an
+``hs-boot`` file, in that it contains only type signatures and type
+declarations, but no implementations::
+
+    -- in Str.hsig
+    signature Str where
+    data Str
+    null   :: Str
+    append :: Str -> Str -> Str
+
+Other modules in the library import this signature and can use
+the types and functions declared in the signature as if it were
+an ordinary module (i.e., *just normal Haskell*, with *no copy-pasting*)::
+
+    -- in Concat.hs
+    module Concat where
+    import Str
+    concat xs = foldr append null xs
+
+Signatures are declared in the Cabal file as a ``required-signature``
+and parametrize *all* of the modules in the library (providing
+*package-level modularity*)::
+
+    library concat-indef
+        required-signatures: Str
+        exposed-modules: Concat
+
+A library with required signatures is called an **indefinite library**.
+As it is missing implementations for its signatures, it cannot be
+compiled; however, it can still be type-checked (*separate
+type-checking*) and registered with the compiler, so that it can be used
+by other indefinite libraries which depend on it.
+
+An indefinite library can be used in two ways:
+
+1. It can be depended upon by other libraries, which inherit
```

Comment:
It can be combined with other libraries to make a compound library that is still indefinite.